mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,516 @@
+    circuit: Circuit,
+    num_training_circuits: int,
+    fraction_non_clifford: float,
+    method_select: str = "random",
+    method_replace: str = "closest",
+    random_state: Optional[Union[int, np.random.RandomState]] = None,
+    **additional_options: dict,
+) -> Tuple[List[Circuit], List[List[float]], List[List[float]]]:
+    """Function to return a list of near-Clifford circuits to create the
+    training data.
+
+    Args:
+        circuit: A circuit of interest.
+        num_training_circuits: Number of circuits in the returned training set,
+                               assumes already compiled into gate set
+                               (Rz, Rx, Z, X, CNOT)
+        fraction_non_clifford: The (approximate) fraction of non-Clifford
+                               gates in each returned circuit.
+        method_select: option to define the way in which the non-Clifford
+                              gates to replace with Cliffords are chosen can
+                              take strings 'random' or 'probabilistic'.
+        method_replace: str = option to define the way the chosen non-Clifford
+                              gates are replace with a Clifford gate can take
+                              strings 'random', 'probabilistic' or 'closest'.
+        random_state: Seed for sampling.
+        additional_options: dictionary with the following keys and values:
+            'sigma_select': float -  postitive variable definined width of
+                                     probability distribution used in choosing
+                                     which non-Cliffords to replace, only has
+                                     an impact if
+                                     method_select = 'probabilistic'.
+            'sigma_replace': float - positive variable definined width of
+                                     probability distribution used in choosing
+                                     which Clifford to replace the
+                                     non-Clifford with, only has an impact if
+                                     method_replace = 'probabilistic'.
+    Returns:
+        List[circ.Circuits]: list of near-Clifford circuits constructed from
+                             the circuits of interest.
+        List[List[float]]: list of list of angles that were replaced in each
+                           training circuit.
+        List[List[float]]: list of list of angles that were inserted in each
+                           training circuit.
+    """
+    circuits_list = []
+    # find all the non-Clifford gates:
+    # all_cliff = np.column_stack((not_rz_circ_data, rz_cliff))
+    angles_original_list = []
+    angles_replaced_list = []
+    # setting the seed:
+    if isinstance(random_state, int):
+        random_state = np.random.RandomState(random_state)
+    else:
+        random_state = randint(10 ** (8))
+        random_state = np.random.RandomState(random_state)
```

Comment:
The function signature says `random_state` can be a `RandomState` but this isn't checked here. If you input a `RandomState` it will not actually use that but use a different one.

Reply:
Yeah thats true thanks for spotting. Are you happy with me just taking an integer argument here for the seed?

Reply:
Yeah this is fine.